File: common/src/main/java/io/github/solusmods/eternalcore/api/storage/StorageEvents.java
package io.github.solusmods.eternalcore.api.storage;

import dev.architectury.event.Event;
import dev.architectury.event.EventFactory;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.chunk.LevelChunk;

import java.util.function.Predicate;

public interface StorageEvents {
    Event<RegisterStorage<Entity>> REGISTER_ENTITY_STORAGE = EventFactory.createLoop();
    Event<RegisterStorage<LevelChunk>> REGISTER_CHUNK_STORAGE = EventFactory.createLoop();
    Event<RegisterStorage<Level>> REGISTER_WORLD_STORAGE = EventFactory.createLoop();

    @FunctionalInterface
    interface RegisterStorage<T extends StorageHolder> {
        void register(StorageRegistry<T> registry);
    }

    @FunctionalInterface
    interface StorageFactory<T extends StorageHolder, S extends AbstractStorage> {
        S create(T target);
    }

    interface StorageRegistry<T extends StorageHolder> {
        <S extends AbstractStorage> StorageKey<S> register(ResourceLocation id, Class<S> storageClass, Predicate<T> attachCheck, StorageFactory<T, S> factory);
    }
}

File: storage-common/src/main/kotlin/io/github/solusmods/eternalcore/storage/api/StorageEvents.kt
package io.github.solusmods.eternalcore.storage.api

import dev.architectury.event.Event
import dev.architectury.event.EventFactory
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.entity.Entity
import net.minecraft.world.level.Level
import net.minecraft.world.level.chunk.LevelChunk
import java.util.function.Predicate

interface StorageEvents {

    fun interface RegisterStorage<T : StorageHolder> {
        fun register(registry: StorageRegistry<T>)
    }

    fun interface StorageFactory<T : StorageHolder, S : Storage> {
        fun create(target: T): S
    }

    interface StorageRegistry<T : StorageHolder> {
        fun <S : Storage> register(
            id: ResourceLocation,
            storageClass: Class<S>,
            attachCheck: Predicate<T>,
            factory: StorageFactory<T, S>
        ): StorageKey<S>
    }

    companion object {
        val REGISTER_ENTITY_STORAGE: Event<RegisterStorage<Entity>> =
            EventFactory.createLoop()

        val REGISTER_CHUNK_STORAGE: Event<RegisterStorage<LevelChunk>> =
            EventFactory.createLoop()

        val REGISTER_WORLD_STORAGE: Event<RegisterStorage<Level>> =
            EventFactory.createLoop()
    }
}